### ISSUE TITLE
Incorrect implementation of template element

### ISSUE THREAD

**Issue by clecap** [tmpvar/jsdom#1295]
In my opinion the <template> element in HTML5 is treated incorrectly by this parser.

According to the html5rocks tutorial as well as the Mozilla implementation, the content of a template element should be placed into its .content and not into its .childNodes.

This deviation from what I understand is the accepted standards currently breaks code which is running in Firefox and Chrome when using jsdom parser.

**Comment by domenic** [comment]
can you provide a small minimal code sample where jsdom behaves differently from browsers?

**Comment by clecap** [comment]
Thanx domenic for your quick and very helpful reply.

TL;DR: Case solved. Issue closed. :-)

Longer: The minimal offending code sample is:

```javascript
   var resu = jsdom.env ("<html><head></head><body><template>Test</template></body></html>",
       function(e,d){
         console.warn (d.document.body.firstChild.content);
   });
```
and it produced "undefined" instead of a DocumentFragment.

However, when re-checking this I noticed that higher up in the code I was not loading jsdom but rather the old node-jsdom. Calling the variable jsdom was not my smartest choice of variable names, as I ended up complaining here instead of at node-jsdom. When I corrected this and updated my package definition the error vanished. 

Thanx for helping me debug a completely different part of my code, I appreciate this, and sorry for taking up your time on this. I think this should now un-break the rest of my code :-) if not I'll be back for more.

**Comment by clecap** [comment]
Sorry, domenic, that I am back with this issue. Now I have a test case and a suggestion wrt the reason behind what I suspect to be a bug in jsdom

**In the browser**, the content inside of a template tag x can be found in the dom at x.content, where it has the form of a document fragment node with the appropriate number of nodes, i.e. in x.content.childNodes (in the example it is 3 childNodes). 

**In jsondom**, the content inside of a template tag x can be found in the dom at x.content, where it first is packaged in a document fragment node with ONE child node; this child node consists of another document fragment node, which then contains the appropriate number of nodes, i.e. in x.content.childNodes.item(0).childNodes. There is an additional, in my opinion unwanted, document fragment wrapper around the content of the template.

Here comes the example.

**In the browser** (firefox) this snippet produces

```
<html><head></head><body><template id="7">a<b>b</b>c</template>
<script>
function domPrintDebugMini (dom) {
  switch (dom.nodeType) {
    case 1:     
      if (dom.tagName == "TEMPLATE") {
        console.warn ("TEMPLATE found: .content has has nodeType: " + dom.content.nodeType + " and contains " + dom.content.childNodes.length + " childNodes; now printing .content");      
        domPrintDebugMini(dom.content);
        console.warn ("DONE printing template");
      }     
      else console.warn (dom.tagName);
      break;
    case 11:
      console.warn ("FRAGMENT found with " + dom.childNodes.length + " childNodes, now printing childNodes");
      for (var i = 0; i < dom.childNodes.length; i++) { domPrintDebugMini (dom.childNodes[i]);}
      console.warn ("DONE printing childNodes");
      break;
  }
}
domPrintDebugMini(document.getElementById("7"));
</script>
</body>
</html>
```
TEMPLATE found: .content has has nodeType: 11 and contains 3 childNodes; now printing .content
mytest.html (line 8)
FRAGMENT found with 3 childNodes, now printing childNodes
mytest.html (line 15)
B
mytest.html (line 12)
DONE printing childNodes
mytest.html (line 17)
DONE printing template
mytest.html (line 10)

**In node.js under jsdom** the following snippet

```
var jsdom          = require ('jsdom');

function domPrintDebugMini (dom) {
  switch (dom.nodeType) {
    case 1:     
      if (dom.tagName == "TEMPLATE") {
        console.warn ("TEMPLATE found: .content has has nodeType: " + dom.content.nodeType + " and contains " + dom.content.childNodes.length + " childNodes; now printing .content");      
        domPrintDebugMini(dom.content);
        console.warn ("DONE printing template");
      }  
      else console.warn (dom.tagName);
      break;
    case 11:
      console.warn ("FRAGMENT found with " + dom.childNodes.length + " childNodes, now printing childNodes");
      for (var i = 0; i < dom.childNodes.length; i++) { domPrintDebugMini (dom.childNodes[i]);}
      console.warn ("DONE printing childNodes");
      break;
  }
}

jsdom.env(
    '<html><head></head><body><template id="7">a<b>b</b>c</template></body></html>',
    function (err, window) {
      domPrintDebugMini (window.document.getElementById("7"));
      console.warn ("DONE");
    }
  );
```
produces


TEMPLATE found: .content has has nodeType: 11 and contains 1 childNodes; now printing .content
FRAGMENT found with 1 childNodes, now printing childNodes
FRAGMENT found with 3 childNodes, now printing childNodes
B
DONE printing childNodes
DONE printing childNodes
DONE printing template
DONE

I suppose that the additional, superfluous document fragment node layer in jsdom is not what is wanted by the standard.

**Comment by domenic** [comment]
Thanks for finding this issue! A pull request with some tests would be much appreciated for correcting it.

**Comment by clecap** [comment]
Sorry, domenic, I do not have a good hand in making pull requests. I am using git only for my own code versioning & backup and probably am getting the references wrong. I did not mean to close this issue here. This was an unintended side effect. Moreover, I do not understand the build system of jsdom to a sufficient degree to run all tests as suggested in the documentation. That said, the code correction which works for me as I expect it (and which has passed the small number of 10 test cases within my own application) can be found at https://github.com/clecap/jsdom/blob/master/lib/jsdom/level2/html.js in line 2282.

I understand that this is an unprofessional way of doing pull requests but I do not want to add more confusion to this thread by yet another misclick.

**Comment by domenic** [comment]
No problem, don't worry about it! When one of us has time we'll do our best to take your changes, add some tests, and pack them up into an official release :)
